### DISCUSSION TITLE
[node] child_process return types too narrow when encoding is unknown.

### DISCUSSION THREAD

**Discussion by codycuellar** [DefinitelyTyped/DefinitelyTyped#55660] [Issues with a @types package] [answered]
The overrides for execSync and execFileSync (possibly others) appears to incorrectly assume Buffer when the options.encoding is not known.
``` typescript
    function execSync(command: string): Buffer;
    function execSync(command: string, options?: ExecSyncOptionsWithStringEncoding): string;
    function execSync(command: string, options?: ExecSyncOptionsWithBufferEncoding): Buffer;
    function execSync(command: string, options?: ExecSyncOptions): Buffer;
```

If the options.encoding is unknown, it's possible the return could be string, but this definition does not allow that and always assumes Buffer. For example:
```typescript
const o: {encoding: 'buffer' | 'utf8'} = {encoding: 'utf8'};
const x = cp.execSync('one', o); // This will be 'string' but the compiler says it can only be Buffer.
```

Am I missing something or is this incorrectly typed? I feel that the last override is simply too narrow and should be:
``` typescript
    function execSync(command: string): Buffer;
    function execSync(command: string, options?: ExecSyncOptionsWithStringEncoding): string;
    function execSync(command: string, options?: ExecSyncOptionsWithBufferEncoding): Buffer;
    function execSync(command: string, options?: ExecSyncOptions): Buffer | string;
```

EDIT:
A better use-case example where the types break:
```typescript
function ls(opts?: cp.ExecSyncOptions) {
    return cp.execSync('ls', opts);
}

const x = ls({encoding: 'utf8'}); // type is Buffer!
```

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
```bash
child_process.execSync('git name-rev --name-only HEAD');
<Buffer 6d 61 73 74 65 72 0a>
child_process.execSync('git name-rev --name-only HEAD', {})
<Buffer 6d 61 73 74 65 72 0a>
child_process.execSync('git name-rev --name-only HEAD', {encoding: 'utf8'})
'master\n'
```
it looks this is correct, what's the runtime code that returns `string` for you?

**Reply by LinusU** [reply]
I think that in the case outlined one doesn't know at _compile-time_ wether or not encoding is `'buffer'` or `'utf8'`, and thus the return value should be typed as `string | Buffer`

**Reply by peterblazejewicz** [reply]
agree, but sample `cp.execSync` does not take undetermined options at all, just action/command

**Reply by codycuellar** [reply]
I realized I messed up the example but I put another example of how wrapping the function produces incorrect types at compile time. The ls function can't be used properly since it SHOULD be returning `string` or at least `string | Buffer`.

**Reply by codycuellar** [reply]
Should I convert this to an issue? I have to do some hacking to make this work correctly as it stands.

**Reply by peterblazejewicz** [reply]
Create draft PR

**Comment by codycuellar** [comment, marked as answer]
I fixed this in [#55835](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55835).
